Add contents entries for domain objects

AA-Turner opened the pull request:
Closes #6316, closes #10804

Per a recent request on the CPython discord, this adds entries in the table of contents for domain objects (e.g. `py:function`, `js:module`, etc).

* [Example page](https://sphinx--10807.org.readthedocs.build/en/10807/extdev/nodes.html)
* [Second example](https://sphinx--10807.org.readthedocs.build/en/10807/extdev/appapi.html)
* [Third example](https://sphinx--10807.org.readthedocs.build/en/10807/extdev/i18n.html)

It does not support objects from the C, C++, and RST domains, as well as the `option`, `cmdoption`, and `envvar` objects.

A [draft implementation](https://gist.github.com/agoose77/e8f0f8f7d7133e73483ca5c2dd7b907f) by @agoose77 and @asmeurer was useful background, thank you both for this.

### Feature or Bugfix
- Feature

### Relates
- https://github.com/python/cpython/issues/83545

Ping for feedback and review, from people who have commented on the linked issues/CPython Discord thread: @Phillip-M-Feldman @slateny @pradyunsg @CAM-Gerlach @tony @agoose77 @asmeurer

A

AA-Turner committed:
Remove `traverse_in_section()`, use `node.findall()`

AA-Turner committed:
Update type signature for `build_toc()`

AA-Turner committed:
Factor out `_make_anchor_name()`

AA-Turner committed:
Add processing for signature description nodes

AA-Turner committed:
Support content in `py:module` and `js:module`

AA-Turner committed:
Add CHANGES entry

AA-Turner committed:
Add tests

AA-Turner committed:
Remove class name from Python methods

AA-Turner committed:
Update test for output format

AA-Turner committed:
Remove `literal` styling

AA-Turner committed:
Remove `literal` styling

asmeurer commented on sphinx/domains/python.py:
```diff
@@ -984,7 +985,14 @@ def run(self) -> List[Node]:
-        ret: List[Node] = []
+
+        content_node: Element = nodes.section()
+        with switch_source_input(self.state, self.content):
```
Glad to see this fixed here too. I didn't look into it too deeply, but should `PyModule` just use the same mechanisms as the other Py* classes (subclassing from `PyObject`)?

Also I suppose the docs should be updated for this.

asmeurer replied:
Was this addressed (documenting the change to `py:module`)?

AA-Turner replied:
Yes:

![image](https://user-images.githubusercontent.com/9087854/189745181-995a6635-29fb-4f62-9f00-cbc70424f891.png)

A

AA-Turner committed:
Update documentation for modules

AA-Turner committed:
Add configuration for ToC qualification control

AA-Turner committed:
Delegate name formatting to domains

AA-Turner committed:
Fix for objects which do not call `ObjectDescription.run()`

AA-Turner committed:
Typo

AA-Turner committed:
Ignore W503

tony commented on tests/test_environment_toctree.py:
```diff
@@ -126,6 +126,46 @@ def test_glob(app):
+
+    # tocs
+    toctree = app.env.tocs['domains']
+    print(toctree[0][1][1][1][0])
```
This and line 164's `print`'s can be cleaned up

AA-Turner committed:
Reinstate `literal` styling

AA-Turner committed:
Merge branch '5.x' into auto-toc

# Conflicts:
#	CHANGES

AA-Turner committed:
Update parent rendering control

- Rename config variable to `toc_object_entries_show_parents`
- Remove `immediate` config value
- Add `domain` config value
- Update documentation
- Update implementations

AA-Turner committed:
Update documentation

AA-Turner committed:
Implement RST domain

AA-Turner committed:
Add the `noindexentry` and `noindex` flags to more domains

AA-Turner committed:
Process entries per signature node

AA-Turner committed:
Indentation

AA-Turner committed:
Update test

AA-Turner committed:
Fix `_object_hierarchy_parts` for invalid input

AA-Turner committed:
Merge branch '5.x' into auto-toc

AA-Turner committed:
Merge branch '5.x' into auto-toc

# Conflicts:
#	CHANGES

AA-Turner committed:
Merge branch '5.x' into auto-toc

AA-Turner committed:
Fix parens dot
